install_sct

```diff
@@ -527,6 +527,11 @@ yes | python/bin/conda create -n venv_sct python=3.6
 # activate miniconda
 source python/etc/profile.d/conda.sh
 conda activate venv_sct
+if [ "$(which python)" != "$(which python/envs/venv_sct/bin/python)" ]; then
```

Comment:
On my Ubuntu machine, `which python/envs/venv_sct/bin/python` still only yields the relative path (hence the CI errors). 

Might be better just to use the explicit path we expect.

```suggestion
if [ "$(which python)" != "$SCT_DIR/python/envs/venv_sct/bin/python" ]; then
```

Reply:
Alternatively, a different check could be:

```suggestion
if [ "$(echo $PATH | cut -f1 -d:)" != "$SCT_DIR/python/envs/venv_sct/bin" ]; then
```

Which covers all the related binaries.

Reply:
Also, I came across an interesting discussion thread on `which`: https://unix.stackexchange.com/questions/85249/why-not-use-which-what-to-use-then

Reply:
Drat. On my Arch machine which gives me absolute paths:

```
[kousu@requiem spinalcordtoolbox]$ which python/envs/venv_sct/bin/python
/home/kousu/src/neuropoly/spinalcordtoolbox/python/envs/venv_sct/bin/python
```

Okay so they suggest `command -v`?

What about readlink or realpath?

```
[kousu@requiem spinalcordtoolbox]$ readlink -f python/envs/venv_sct/bin/python
/home/kousu/src/neuropoly/spinalcordtoolbox/python/envs/venv_sct/bin/python3.6
```

```
[kousu@requiem spinalcordtoolbox]$ realpath python/envs/venv_sct/bin/python
/home/kousu/src/neuropoly/spinalcordtoolbox/python/envs/venv_sct/bin/python3.6
```

but that, as in this example, might still not be quite right because "3.6" got appended there.

Oh but running the same on both sides should work, shouldn't it?

```suggestion
[kousu@requiem spinalcordtoolbox]$ if [ "$(command -v /usr/bin/python)" = "$(command -v python3)" ]; then echo Aye; fi
[kousu@requiem spinalcordtoolbox]$ 
```

Hm nope. Why? Even though these are the same program, `command` doesn't resolve symlinks:

```
[kousu@requiem spinalcordtoolbox]$ command -v python
/usr/bin/python
[kousu@requiem spinalcordtoolbox]$ command -v python3
/usr/bin/python3
[kousu@requiem spinalcordtoolbox]$ ls -l $(command -v python)
lrwxrwxrwx 1 root root 7 Dec 13 06:55 /usr/bin/python -> python3
[kousu@requiem spinalcordtoolbox]$ ls -l $(command -v python3)
lrwxrwxrwx 1 root root 9 Dec 13 06:55 /usr/bin/python3 -> python3.9
```

Both, then?

```
[kousu@requiem spinalcordtoolbox]$ if [ "$(readlink -f "$(command -v /usr/bin/python)")" = "$(readlink -f "$(command -v python3)")" ]; then echo Aye; fi
Aye
```

Or, I guess we don't really need to worry about symlinks? We just want to know we're pointed at the right subfolder.

Reply:
Tried one of them. Gonna be lazy and wait for CI to see if it worked.

Reply:
Bleh not yet, okay I'll take another swipe at this.

Reply:
Okay the problem with `command -v` is the same as with `which`: it doesn't get the absolute path necessarily:

```
$ command -v python
/home/kousu/src/neuropoly/spinalcordtoolbox/python/envs/venv_sct/bin/python
$ command -v python/envs/venv_sct/bin/python
python/envs/venv_sct/bin/python
```

Reply:
Okay I solved it by stacking `realpath` on `command -v`. Here's what the check looks like on my machine:

```
+ '[' /home/kousu/src/neuropoly/spinalcordtoolbox/python/envs/venv_sct/bin/python3.6 '!=' /home/kousu/src/neuropoly/spinalcordtoolbox/python/envs/venv_sct/bin/python3.6 ']'
```

Unfortunately, `realpath` doesn't exist on macOS:

- https://stackoverflow.com/questions/1055671/how-can-i-get-the-behavior-of-gnus-readlink-f-on-a-mac
- macOS's libc *has* `realpath()`: https://developer.apple.com/library/archive/documentation/System/Conceptual/ManPages_iPhoneOS/man3/realpath.3.html; but it doesn't exist expose it; someone wrote a whole C program just to expose it: https://github.com/harto/realpath-osx/blob/master/realpath.c

Reply:
Ah, whoops, I think there might have been a miscommunication on my part. Sorry about that. :sweat_smile: 

1. `command -v python` give the absolute path of `python`.
2. `command -v python/envs/venv_sct/bin/python` is the one that gives the relative path.

So, 1. is fine to leave as-is. For 2., since we already know the full path that we're expect, can't we just use the filepath as-is? (See https://github.com/neuropoly/spinalcordtoolbox/pull/3103#discussion_r545236719) That way, no `realpath` shenanigans are needed, I think.

Reply:
No apologies needed. Really!

My thinking is that I want something that reliably gives the same format. Neither `which` nor `command` do that; because their algorithms are basically: try concatenating to each thing in $PATH until something catches and output. Plus symlinks can get involved (they often do: `$path/$to/python` is usually linked to `$path/$to/python3` is usually linked to `$path/$to/python3.7`) and that makes comparisons more unreliable.

Reply:
Well, even if symlinks are involved, won't `$path/$to/python`, `$path/$to/python3`, and `$path/$to/python2.7` all be `!=` to the Miniconda python if something's amiss? Even if we don't have the perfectly accurate python interpreter on the left side, we're only really checking 1 possible path on the right side. :yum:

Reply:
I dunno. Maybe you're right. But in my experience there's too many ways you can get tripped up by thinking you know how things work and that they'll work in all cases and then they don't. [realpath(3)](https://www.man7.org/linux/man-pages/man3/realpath.3.html) is guaranteed to give a canonical answer, so we'll know.

The only way I can think it might fail is if hardlinks are involved too, but then I'd debate even more...

Reply:
OK, I'm sold. :slightly_smiling_face:

Reply:
Looks like you were on the nose with this @joshuacwnewton: `which` is being **deleted**: https://lwn.net/SubscriberLink/874049/bf89a969ed3dde87/, https://lwn.net/ml/debian-devel/9e3a2df146f73437fe8cce10272ddf3913e48757.camel@debian.org/

Reply:
Oh my!